### DISCUSSION TITLE
Clarification about bytecode generation during cross-compilation

### DISCUSSION THREAD

**Discussion by mattiapv** [yao-pkg/pkg#94] [Q&A] [answered]
I've been using `pkg` for a while, but I never fully understood if bytecode is generated for all architectures.
From the README:

> If you want to generate an executable for different architectures, note that by default, pkg needs to run the executable of the target architecture to generate bytecodes.

What does this mean exactly? Does it imply that if I want to generate bytecode for Windows, I need to run `pkg` on a Windows machine, not on macOS? 

I tried unpacking the `.exe` file generated from macOS, but I couldn’t read my code. However, when I used the `--no-bytecode` flag, I was able to read the source code.

When I asked in another repository ([bytenode](https://github.com/bytenode/bytenode)) whether bytecode for Windows could be generated on macOS, I got this response:

> This is not possible. Not in bytenode, pkg, or any other libraries. Bytecodes must be generated on the exact same architecture that will run them in production.

So, is pkg able to generate bytecode for all architectures from a single machine?
Thanks in advance for any clarification

**Comment by robertsLando** [comment, marked as answer]
pkg do generate the bytecode of you project for each target you specify (node-os-arch), what's changing then is the nodejs binary we bundle the bytecode to that of course must match the target  os/arch but of course we do support cross os and arch builds

The bytecode production is done here: https://github.com/yao-pkg/pkg/blob/main/lib/fabricator.ts

**Reply by mattiapv** [reply]
Thank you for the quick response. There's another question I haven't found an answer to: why does the Node.js binary need to be patched every time a new release comes out in order to work with `pkg`?

**Reply by robertsLando** [reply]
Them allow to make node loading things from pkg snapshot fs and use bytecode from snapshot directly instead of tyring to convert the code to v8
